go/chat/ephemeral_purger.go

```diff
@@ -0,0 +1,315 @@
+package chat
+
+import (
+	"container/heap"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/chat/globals"
+	"github.com/keybase/client/go/chat/storage"
+	"github.com/keybase/client/go/chat/types"
+	"github.com/keybase/client/go/chat/utils"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/client/go/protocol/gregor1"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/clockwork"
+)
+
+// An queueItem is something we manage in a priority queue.
+type queueItem struct {
+	purgeInfo chat1.EphemeralPurgeInfo
+
+	// The index is needed by `update` and is maintained by the heap.Interface methods.
+	index int // The index of the queueItem in the heap.
+}
+
+// A priorityQueue implements heap.Interface and holds queueItems.
+// We also keep a map of queueItems for easy item updates
+type priorityQueue struct {
+	queue []*queueItem
+	// convID -> *queueItem
+	itemMap map[string]*queueItem
+}
+
+func newPriorityQueue() *priorityQueue {
+	return &priorityQueue{
+		queue:   []*queueItem{},
+		itemMap: make(map[string]*queueItem),
+	}
+}
+
+func (pq priorityQueue) Len() int { return len(pq.queue) }
+
+func (pq priorityQueue) Less(i, j int) bool {
+	return pq.queue[i].purgeInfo.NextPurgeTime < pq.queue[j].purgeInfo.NextPurgeTime
+}
+
+func (pq priorityQueue) Swap(i, j int) {
+	pq.queue[i], pq.queue[j] = pq.queue[j], pq.queue[i]
+	pq.queue[i].index = i
+	pq.queue[j].index = j
+}
+
+func (pq *priorityQueue) Push(x interface{}) {
+	item := x.(*queueItem)
+	item.index = len(pq.queue)
+	pq.queue = append(pq.queue, item)
+	pq.itemMap[item.purgeInfo.ConvID.String()] = item
+}
```

Comment:
Do we want to check for callers doing `Push` on the same convo twice?

Reply:
I left a comment to not do that and use `update` directly. it wouldn't break things necessarily if someone used this but it's not the intended behavior